distributed/minGPT-ddp/mingpt/trainer.py

```diff
@@ -0,0 +1,140 @@
+"""
+Simple training loop; Boilerplate that could apply to any arbitrary neural network,
+so nothing in this file really has anything to do with GPT specifically.
+"""
+
+from dataclasses import dataclass, asdict
+from collections import OrderedDict
+from typing import Optional, Any, Dict
+import os
+
+import torch
+from torch.utils.data import Dataset, DataLoader
+from torch.nn.parallel import DistributedDataParallel as DDP
+from torch.utils.data.distributed import DistributedSampler
+
+import boto3
+from urllib.parse import urlparse
+import fsspec
+import io
+
+@dataclass
+class TrainerConfig:
+    max_epochs: int = None
+    batch_size: int = None
+    data_loader_workers: int = None
+    grad_norm_clip: float = None
+    snapshot_path: Optional[str] = None
+    save_every: int = None
+
+@dataclass
+class Snapshot:
+    model_state: 'OrderedDict[str, torch.Tensor]'
+    optimizer_state: Dict[str, Any]
+    finished_epoch: int
+
+def upload_to_s3(obj, dst):
+    buffer = io.BytesIO()
+    torch.save(obj, buffer)
+    buffer.seek(0)
+    dst = urlparse(dst, allow_fragments=False)
+    boto3.client('s3').upload_fileobj(buffer, dst.netloc, dst.path.lstrip('/'))
+
+class Trainer:
+
+    def __init__(self, trainer_config: TrainerConfig, model, optimizer, train_dataset, test_dataset=None):
+        self.config = trainer_config
+        # set torchrun variables
+        self.local_rank = int(os.environ["LOCAL_RANK"])
+        self.global_rank = int(os.environ["RANK"])  
+        # data stuff
+        self.train_dataset = train_dataset
+        self.train_loader = self._prepare_dataloader(train_dataset)
+        self.test_loader = self._prepare_dataloader(test_dataset) if test_dataset else None
+        # initialize train states
+        self.epochs_run = 0
+        self.model = model.to(self.local_rank)
+        self.optimizer = optimizer        
+        self.save_every = self.config.save_every
+        # load snapshot if available. only necessary on the first node.
+        if self.config.snapshot_path is None:
+            self.config.snapshot_path = "snapshot.pt"
+        self._load_snapshot()
+        # wrap with DDP. this step will synch model across all the processes.
+        self.model = DDP(self.model, device_ids=[self.local_rank])
+        
+    def _prepare_dataloader(self, dataset: Dataset):
+        return DataLoader(
+            dataset,
+            batch_size=self.config.batch_size,
+            pin_memory=True,
+            shuffle=False,
+            num_workers=self.config.data_loader_workers,
+            sampler=DistributedSampler(dataset)
+        )
+
+    def _load_snapshot(self):
+        try:
+            snapshot = fsspec.open(self.config.snapshot_path)
```

Comment:
Since there's a lot of snapshot related code here, curious if you've tried https://github.com/pytorch/torchsnapshot and what you thought @yifuwang @ananthsub

Reply:
I haven't! I'd like to update the code with torchsnapshot once this gets merged in